### ISSUE TITLE
GCC version tracking

### ISSUE THREAD

**Issue by tonytheodore** [mxe/mxe#1205]
Following on from #965 and https://github.com/mxe/mxe/commit/eae41ccbfd983312b99cbf1461d6217b701a57b0.

@avih wrote:
> MXE has always been about being at the front with regards to packages versions used

Another way of looking at it is that MXE strives to make it possible/easy to keep updated, actually being at the front is more an artefact and not a primary goal (see far below).

> Whoever needs/wants the remaining fix would still be able to use gcc4 as a plugin

From a users perspective, I'm not sure how different that is from it's inverse of using a plugin for the new version ;)

> But gcc5 is where all the progress happens 

GCC has a [new versioning scheme](https://gcc.gnu.org/develop.html#num_scheme) and v6 has just entered it's [fourth and final stage](https://gcc.gnu.org/ml/gcc/2016-01/msg00168.html). New features are waiting for v7.

> many distros use it by default these days

We can't use the major distros as a proxy for stability - they have very different criteria and orders of magnitude more packages, low-level subsystems, and users to consider. They are also primary platforms for [gcc release criteria](https://gcc.gnu.org/gcc-5/criteria.html) so any given version is likely to be more stable than it's mingw equivalent.

Of course, the corollary is that MXE has a much smaller set of packages and doesn't need to build an entire operating system. This greatly reduces requirements in terms of stability - if the distros are using it, we are very far behind.

> I'm sure gcc4 has bugs too.

Of course, both [early general issues](http://lkml.iu.edu//hypermail/linux/kernel/1407.3/00650.html) and [current MXE ones](https://github.com/mxe/mxe/issues/1184)

@starius wrote:
> Do MXE users use features introduced after gcc4.9?

[Yes](https://github.com/mxe/mxe/pull/965#issuecomment-157351748), though there's  perennial interest in new c++ standards. Smaller numbers of users at this stage in the cycle is more an argument in favour of updating to get past the chicken-and-egg problem.

See also @mabrand's https://github.com/mxe/mxe/pull/965#issuecomment-155354642 - I'm not sure we have much external or objective criteria to base decisions on.

I find it useful to think of MXE as three things: build infrastructure, compiler toolchain, and a set of packages. To improve any area, it's easier when the other two are relatively static. We have some nice infrastructure improvements recently (mostly `build-pkg`) and all packages are building and progress is being made with updates - it's a good time to think about the compiler.

I wonder if the question of gcc versions is more a matter of not having sufficient infrastructure to have more granular control. For example, many gcc bugs are triggered by a specific set of flags and we could use later versions with those flags configured in a certain way by default.

The new version scheme means we'll be talking about a gcc6 release in 3 months. It doesn't mean what it used to, but historically we likely would have updated very quickly.

At the moment, I think it makes the most sense for MXE to align updates with the new gcc schedule and make it easier for people to fall back to previous versions. This should have the following benefits:
 * allows users access to new features and fixes earlier
 * allows conservative users to update at their own pace
 * expose and fix bugs to benefit the general mingw ecosystem and future gcc releases

The last point is probably the most contentious as it presumes resources we don't have - it's also the most important. I have referenced `mingw` but it's really `mingw-w64` these days - smaller projects need to balance relative stability with future relevance (don't mean to be overly dramatic ;).

Before gcc6 lands, I hope we have a better basis for such decisions.

**Comment by avih** [comment]
> From a users perspective, I'm not sure how different that is from it's inverse of using a plugin for the new version ;)

Sure you do. Default counts. Without any data, I'd bet the vast majority of MXE users use the default config, as users do everywhere else.

**Comment by avih** [comment]
Ping? :) (change the default `gcc` to something newer)

**Comment by starius** [comment]
See https://github.com/mxe/mxe/pull/1541

**Comment by avih** [comment]
Not 6.x ?

**Comment by starius** [comment]
I want to go to gcc6 with a stop in gcc5. Switch to gcc6 seems to require more patching and it is better to be on gcc5 not on gcc4 at that time.

**Comment by avih** [comment]
> Switch to gcc6 seems to require more patching and it is better to be on gcc5 not on gcc4 at that time.

Agreed.

**Comment by tonytheodore** [comment]
> I wonder if the question of gcc versions is more a matter of not having sufficient infrastructure to have more granular control. For example, many gcc bugs are triggered by a specific set of flags and we could use later versions with those flags configured in a certain way by default.

I now think this is entirely the case.

> I find it useful to think of MXE as three things: build infrastructure, compiler toolchain[...]

> At the moment, I think it makes the most sense for MXE to align updates with the new gcc schedule and make it easier for people to fall back to previous versions.


I now think it makes more sense for mxe to have no specific default and make the compiler an explicit part of the target. We can't know when people are ready to upgrade or revert - we can make it easy to do those things in parallel.

Note I say compiler, but I really mean toolchain. We have the "cc" meta-package which also enables side-by-side `clang` and `glibc` if we use explicit targets and move away from defaults.

**Comment by avih** [comment]
> I now think it makes more sense for mxe to have no specific default and make the compiler an explicit part of the target

What does that mean? I don't think it means that each package will be configured to use a different version of gcc, right?

**Comment by tonytheodore** [comment]
A better way to say it is that mxe will have no "implied" default compiler, when a user tries to build a target like `i686-w64-mingw32.static` they will get a message like:
```
Targets without a specified compiler are deprecated, please use:
i686-w64-mingw32.static.gcc[5-10]
```
Once they have adjusted any scripts to use target variables (likely already done for i686*,x86_64*), it is then trivial to test newer compilers at a pace that makes sense for their project. They can do this side-by-side, having release and beta builds etc.

Essentially, people should be able to explicitly choose the compiler instead of us trying to balance the different use cases.

**Comment by avih** [comment]
> instead of us trying to balance the different use cases.

Not sure about that actually. Why would I know better than you what compiler I should choose?
